### PR TITLE
[bootstrap] Add support for building gcc and libgccjit

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
As @eholk summarized below:

> From my understanding, this change would add libgccjit as an optional component to the Rust distribution. This library is licensed under GPLv2 and currently we do not have any other components under that license so it would be a new license, and one that is generally more restrictive than the other licenses we use.

It'll greatly improve the experience for anyone wanting to work on the GCC backend from the compiler.

Should help with https://github.com/rust-lang/rust/issues/124172.
Will unblock #124353.

r? @Kobzol

### CODE REVIEW

**Comment by GuillaumeGomez** [comment]
From what I can see, it's not possible to tell `reuse` to allow deprecated licenses. So in this case, I have no clue how we can go around this issue. If someone has an idea?

**Comment by Kobzol** [comment]
@davidtwco Any suggestions on what to do about the GPL-2.0 license of GCC?

**Comment by davidtwco** [comment]
> @davidtwco Any suggestions on what to do about the GPL-2.0 license of GCC?

Will get back to you about this

**Comment by pietroalbini** [comment]
> From what I can see, it's not possible to tell reuse to allow deprecated licenses. So in this case, I have no clue how we can go around this issue. If someone has an idea?

Note that REUSE doesn't mark the GPL 2.0 license as deprecated. What is deprecated is referring to it as `GPL-2.0`. You should refer to the license as either `GPL-2.0-only` or `GPL-2.0-or-later`, depending on what is appropriate.

Also, the warning doesn't come from REUSE itself, but from the [SPDX License List](https://spdx.org/licenses/) (you'll find `GPL-2.0` as deprecated in there).

**Comment by eholk** [comment]
We discussed this at the Council meeting today and we don't have any concerns with this change. However, we would like to confirm with the Foundation. @abibroom - do you know of any concerns from the Foundation side (feel free to ping me on Zulip or reach out via email if you'd rather not discuss this publicly.

From my understanding, this change would add libgccjit as an optional component to the Rust distribution. This library is licensed under GPLv2 and currently we do not have any other components under that license so it would be a new license, and one that is generally more restrictive than the other licenses we use.

File: src/bootstrap/src/core/build_steps/dist.rs
Comment (Kobzol) [inline src/bootstrap/src/core/build_steps/dist.rs:2369]: What will this do on Windows? :thinking: Does the gcc build also work on non-Linux OSes, in general?

**Comment by Kobzol** [comment]
@bors try

Just to see what happens on Linux dist.

**Comment by nikic** [comment]
Can this be done without adding gcc as a submodule?

**Comment by abibroom** [comment]
> We discussed this at the Council meeting today and we don't have any concerns with this change. However, we would like to confirm with the Foundation. @abibroom - do you know of any concerns from the Foundation side (feel free to ping me on Zulip or reach out via email if you'd rather not discuss this publicly.

I'd like to be able to say this is all fine, but want to run it past Foundation legal counsel first just to make sure. Please bear with me whilst that happens!

**Comment by eholk** [comment]
> Enabling downstream compliance: "When it comes to downstream distribution, anyone providing a binary distribution of Rust incorporating this component will need to provide source code for the whole distribution. This is ultimately their problem, but if you're aware of anyone who's doing this currently, it's probably worth talking through those cases and figuring out how to provide notice to them and other potential distributors of their additional obligations with respect to this component."

@abibroom - just to make sure, if someone wanted to make a downstream binary distribution without releasing source code (which strikes me as rather unlikely), they could still do so if they chose not to include the libgccjit component?

I want to make sure this change doesn't effectively mean we have to relicense Rust as GPL2.

**Comment by GuillaumeGomez** [comment]
Unless you compile with the GCC backend, the license of GCC doesn't concern you.

**Comment by Kobzol** [comment]
> Unless you compile with the GCC backend, the license of GCC doesn't concern you.

It would be great if this was clarified a bit more in the discussion. When we were talking about "binary distributions" here, it means a distribution of the compiler toolchain, right? Not the actual end user Rust code compiled with rustc.

**Comment by GuillaumeGomez** [comment]
This is the part I'm not clear about: by default, do we plan to include the GCC backend in all binary distributions? `libccjit.so` and `gcc` are only used at compile time. Using them changes the license of the generated binaries. However, does having them in the binary distribution also implies that all other items are impacted by the license?

**Comment by eholk** [comment]
> libccjit.so and gcc are only used at compile time. Using them changes the license of the generated binaries.

Are you saying that if you compile code with gcc then the resulting binaries have to be GPL? As far as I know, that's not the case. I think it's relatively common for closed-source products to be compiled and released using GPL toolchains.

Generally I thought the way it works is if you include GPL code in your compiled binary then your whole binary needs to be GPL.

**Comment by GuillaumeGomez** [comment]
That's not how I understood it. But I'm definitely not an expert in the area so I'd better keep quiet until someone well versed in licenses and in GPL2 in particular could confirm our thoughts.

**Comment by abibroom** [comment]
> @abibroom - just to make sure, if someone wanted to make a downstream binary distribution without releasing source code (which strikes me as rather unlikely), they could still do so if they chose not to include the libgccjit component?

Yes.

> When we were talking about "binary distributions" here, it means a distribution of the compiler toolchain, right? Not the actual end user Rust code compiled with rustc.

Yes; "a binary distribution **of Rust**" where 'Rust' means what is provided in this GitHub org/repo by the Rust Project.

**Comment by bjorn3** [comment]
Assuming that cg_gcc does the same for distribution as cg_clif, all GCC components and the GCC backend itself will end up in a rustc-codegen-gcc-preview component which would need to be separately installed (using eg `rustup component add rustc-codegen-gcc-preview`) to use the GCC backend.

File: src/bootstrap/src/core/build_steps/dist.rs
Comment (GuillaumeGomez) [reply src/bootstrap/src/core/build_steps/dist.rs:2369]: No it doesn't, good catch!

**Comment by GuillaumeGomez** [comment]
> > From what I can see, it's not possible to tell reuse to allow deprecated licenses. So in this case, I have no clue how we can go around this issue. If someone has an idea?
> 
> Note that REUSE doesn't mark the GPL 2.0 license as deprecated. What is deprecated is referring to it as `GPL-2.0`. You should refer to the license as either `GPL-2.0-only` or `GPL-2.0-or-later`, depending on what is appropriate.
> 
> Also, the warning doesn't come from REUSE itself, but from the [SPDX License List](https://spdx.org/licenses/) (you'll find `GPL-2.0` as deprecated in there).

As you can see @pietroalbini, REUSE wants `GPL-2.0` specifically, any other will not work. And SPDX is considering GPL-2.0 as deprecated. So unless I'm missing something, we're stuck here. Any idea?

**Comment by Skgland** [comment]
Based on https://github.com/rust-lang/rust/pull/125419#issuecomment-2126672496 it looks like the deprecated license originates in three files 

> 'GPL-2.0' found in:
> * ../src/gcc/gcc/testsuite/c-c++-common/analyzer/asm-x86-dyndbg-2.c
> * ../src/gcc/gcc/testsuite/c-c++-common/analyzer/taint-assert-BUG_ON.c
> * ../src/gcc/gcc/testsuite/gcc.dg/analyzer/asm-x86-dyndbg-1.c

would it be possible to patch these to specify `GPL-2.0-only` or `GPL-2.0-or-later` instead?

**Comment by GuillaumeGomez** [comment]
I'll simply remove them...

**Comment by Skgland** [comment]
Based on https://github.com/torvalds/linux/blob/master/LICENSES/preferred/GPL-2.0
```
For 'GNU General Public License (GPL) version 2 only' use:
    SPDX-License-Identifier: GPL-2.0
  or
    SPDX-License-Identifier: GPL-2.0-only
```
and the comment mentioning that the license is chosen based on them being derived from Linux kernel files I think `GPL-2.0-only` should be considered equivalent

**Comment by davidmalcolm** [comment]
> As @eholk summarized below:
> 
> > From my understanding, this change would add libgccjit as an optional component to the Rust distribution. This library is licensed under GPLv2 and currently we do not have any other components under that license so it would be a new license, and one that is generally more restrictive than the other licenses we use.

Note that the libgccjit library is licensed under GPLv3, not GPLv2 (I'm the author/maintainer of libgccjit, but it's mostly just a thin wrapper around GCC's own code hence GPLv3).

Are you basing the GPLv2 thing on those SPDX identifiers from the three files in the testsuite?  Those are taken/adapted from the Linux kernel and thus GPLv2 only, AIUI.

**Review by antoyo — APPROVED** [review]

**Comment by nikic** [comment]
I'd like to repeat this question:

> Can this be done without adding gcc as a submodule?

Does distributing libgcc require gcc to be an actual submodule in the rust repo?

Does this make gcc part of the rustc-src tarballs?

**Comment by GuillaumeGomez** [comment]
> I'd like to repeat this question:
> 
> > Can this be done without adding gcc as a submodule?
> 
> Does distributing libgcc require gcc to be an actual submodule in the rust repo?
> 
> Does this make gcc part of the rustc-src tarballs?

I originally went for a commit hash stored in a file but I was asked (at least that's what I remember but maybe I'm wrong) to use a git submodule instead to make things simpler.

> Does this make gcc part of the rustc-src tarballs?

Normally no but we'll need someone from infra to answer this question.

**Comment by Kobzol** [comment]
> Does this make gcc part of the rustc-src tarballs?

Yes, because the submodule is under the `src` directory, and everything there is included by [default](https://github.com/rust-lang/rust/blob/59681be64731a290b13d36e959534d6f5672b76d/src/bootstrap/src/core/build_steps/dist.rs#L989).

As of the current shape of this PR, there is `src/gcc` with all the GCC glory in our source tarballs.

**Comment by GuillaumeGomez** [comment]
Thanks for clarifying @Kobzol !

So what remains to be discussed here?

**Comment by Kobzol** [comment]
@nikic Do you have concerns about us distributing GCC source code?

(it feels like we should have `t-legal` or `t-licensing` :) )

**Comment by wesleywiser** [comment]
Not a lawyer and ultimately not my call here but what would make the most sense to me is that cg_gcc is its own rustup component. If you do not explicitly ask for it, it is not installed _and_ none of the GPL'd source code is present in any other component. If you ask for cg_gcc to be installed you get both the binaries and the source for them at the same time. (Note, since `compiler/rustc_codegen_gcc` is itself Apache 2/MIT licensed, I think it's fine if that code is present in the `rustc-src` package but there should not be any `GCC` source code there.)

I believe that fully satisfies license requirements to make the source available with binaries and it also clearly separates the GPL code from the rest of the components. 

What we want to avoid is creating a perception that the compiler depends on GPL'd code in a way that would trigger the copyleft portions of the license.

**Comment by GuillaumeGomez** [comment]
I thought it was the case. If not, how can I make it into its own component?

**Comment by nikic** [comment]
> @nikic Do you have concerns about us distributing GCC source code?
> 
> (it feels like we should have `t-legal` or `t-licensing` :) )

Basically what @wesleywiser said. I think this PR is fine from a *legal* perspective, but it gives an incorrect *impression* of rustc's licensing situation. We now have gcc code in the repo, and in the rustc-src tarball, and GPL licenses inside the LICENSES directory. A human would likely jump to the wrong conclusions about what this means, and any automated license analysis tooling certainly would.

**Comment by Kobzol** [comment]
Maybe we could remove the GCC sources from the main rustc-src tarball and only distribute them in the GCC component?

**Comment by GuillaumeGomez** [comment]
Sounds good to me. Should I checkout GCC from a commit stored in a file then?

**Comment by Kobzol** [comment]
I personally don't think that's necessary, but IANAL, of course. I'd just avoid including GCC source code and its GPL license(s) in the Rust toolchain source code tarballs. This would require some modifications to the `rustc-src` bootstrap step and potentially also the `collect-license-metadata` and `generate-copyright` tools.

File: src/bootstrap/src/core/build_steps/dist.rs
Comment (Kobzol) [inline src/bootstrap/src/core/build_steps/dist.rs:2289]: Let's move the Linux x64 check out of `Gcc` and put it here. Both to make it more explicit that we're only packaging `Gcc` in a selected platform, and also to make it possible to execute the `Gcc` step manually, perhaps even on different platforms.

File: src/bootstrap/src/core/build_steps/gcc.rs
Comment (Kobzol) [inline src/bootstrap/src/core/build_steps/gcc.rs:40]: Having both `update_existing_submodule` and `update_submodule` is not needed, the second call is enough.

File: src/bootstrap/src/core/build_steps/gcc.rs
Comment (Kobzol) [inline src/bootstrap/src/core/build_steps/gcc.rs:81]: Let's remove the commented code for now, and just focus on actually building `gcc` in this PR.

File: src/bootstrap/src/core/build_steps/gcc.rs
Comment (Kobzol) [inline src/bootstrap/src/core/build_steps/gcc.rs:184]: There is no `download-ci-gcc` yet, this comment is thus outdated (or maybe future-dated :) ).

File: src/bootstrap/src/core/build_steps/llvm.rs
Comment (Kobzol) [inline src/bootstrap/src/core/build_steps/llvm.rs:1144]: Could you move this e.g. to `src/utils/helpers.rs` instead of exporting it from the LLVM module?

**Review by Kobzol — COMMENTED** [review]
I would first focus only on the step of building GCC (and libgccjit), and keep distribution for follow-up PRs. It would be useful to add a `x build` step for `gcc`, so that it can be tested locally that it can be indeed built.

File: src/bootstrap/src/core/build_steps/gcc.rs
Comment (GuillaumeGomez) [reply src/bootstrap/src/core/build_steps/gcc.rs]: Future-dated for sure. I was having sight maybe too far from present time. :sweat_smile:

File: src/bootstrap/src/core/build_steps/llvm.rs
Comment (GuillaumeGomez) [reply src/bootstrap/src/core/build_steps/llvm.rs]: Sure, done!

File: .gitmodules
Comment (GuillaumeGomez) [reply .gitmodules]: Done!

File: src/bootstrap/src/core/build_steps/gcc.rs
Comment (GuillaumeGomez) [reply src/bootstrap/src/core/build_steps/gcc.rs]: Done.

File: src/bootstrap/src/core/build_steps/dist.rs
Comment (GuillaumeGomez) [reply src/bootstrap/src/core/build_steps/dist.rs]: Done as well.

**Comment by GuillaumeGomez** [comment]
Not sure why it's complaining that we have unused licenses now though...

**Comment by Kobzol** [comment]
Could you please add `gcc::Gcc` [here](https://github.com/rust-lang/rust/blob/486864f23593d7aedddf1831f748660311f9ff3c/src/bootstrap/src/core/builder.rs#L796), so that it is possible to do `./x build gcc`?

The licenses are not found because we didn't tell REUSE yet how to apply them. We need something like [this](https://github.com/rust-lang/rust/blob/37ab09010cb627df6b3ffb2d4e95c2cacaf93efb/REUSE.toml#L171-L170).

File: .gitmodules
Comment (Kobzol) [inline .gitmodules:52]: Could you also please add `shallow = true` here? Otherwise it's very.. heavy to clone :laughing:

Comment (GuillaumeGomez) [reply .gitmodules:52]: Yes it is. ^^'

File: src/bootstrap/src/core/build_steps/gcc.rs
Comment (Kobzol) [inline src/bootstrap/src/core/build_steps/gcc.rs:44]: Ideally this should be `builder.src.join("src/gcc")`, as the current working directory might not always be the source root (although that's rare).

Comment (GuillaumeGomez) [reply src/bootstrap/src/core/build_steps/gcc.rs]: Much better indeed! :+1:

**Comment by Kobzol** [comment]
```
configure: error: Building GCC requires GMP 4.2+, MPFR 3.1.0+ and MPC 0.8.0+.
Try the --with-gmp, --with-mpfr and/or --with-mpc options to specify
their locations.  Source code for these libraries can be found at
their respective hosting sites as well as at
```
I don't suppose there's a way to tell GCC to build these for us? I'm not sure if we require any special external dependencies for LLVM.

**Comment by antoyo** [comment]
@Kobzol: I do not know where you see this error, though.

I see a licensing error:

> Unfortunately, your project is not compliant with version 3.2 of the REUSE Specification :-(

Do you have any idea how to fix this?

**Comment by Kobzol** [comment]
I saw the error when I tried to build the GCC bootstrap step locally. We should ideally make the build selfcontained, by either running the download dependencies script in bootstrap or just preparing the various submodules beforehand in bootstrap, so that GCC builds them automatically.

The license error is caused by reuse not detecting the GCC licenses added in https://github.com/rust-lang/rust/pull/125419/files#diff-74bd2d6afd7a1f967679adae374ca38fd9745fe6129dc4717a1ab0e388590b08R172 for some reason. I haven't investigated why yet.

**Comment by GuillaumeGomez** [comment]
Your help would be very appreciated! I updated to new bootstrap API and fixed the merge conflict to make it easier for you.

**Comment by Kobzol** [comment]
I pushed a fix for the licensing stuff (hopefully), there were some small issues in the `REUSE.toml` file (invalid path and wrong order of the directory overrides). I still want to make sure that it is easy to build GCC locally (and also in CI), by downloading the dependencies in bootstrap. I will try to integrate the dependency installation into bootstrap tomorrow.

**Comment by GuillaumeGomez** [comment]
Hopefully, you should never need to build it locally (although for CI issues, better for reproduciability). Seems like you fixed the CI, thanks a lot!

**Comment by Kobzol** [comment]
I thought that one of the goals of this PR is to make it easier for developers to build and contribute to rustc_codegen_gcc :thinking: So I assumed that we want to also automate building of GCC as much as possible.

**Comment by Kobzol** [comment]
But it's true that probably most users will just want to download libgccjit from Ci, same as they do for LLVM.

**Comment by antoyo** [comment]
The endgoal is to have rustup distribution of rustc_codegen_gcc, but having it easy to run locally would be helpful for some devs indeed. Please tell me if you get any error while doing so.

I'll also try on my end to run that locally.

**Comment by Kobzol** [comment]
I used the `contrib` script to download dependencies and it seems like GCC is being built successfully. But it is looooooong. If we are to enable this on CI, we'll either need to set it up with sccache, or perhaps even completely prebuilt it in Docker, if that doesn't help enough.

**Comment by antoyo** [comment]
How did you build it? Adding `-j24` to `make` will help a lot.

**Comment by Kobzol** [comment]
Not on a laptop with `8` cores xD It uses `-j<bootstrap-cores>`, so it should have been `-j8`. But this is not that important right now, we'll see later how it performs on CI.

**Comment by antoyo** [comment]
Anything else needed before this can be approved and merged?
Thanks for the license fix, btw.

File: src/bootstrap/src/core/build_steps/gcc.rs
Comment (Kobzol) [inline src/bootstrap/src/core/build_steps/gcc.rs:127]: Do we need this? On a local build, it failed with an error, because the target file (symlink) has already existed. We could guard this with a condition that would not create the symlink if it already exists, but if the GCC build already creates the symlink, then maybe we don't need this code at all?

File: src/bootstrap/src/core/build_steps/gcc.rs
Comment (Kobzol) [inline src/bootstrap/src/core/build_steps/gcc.rs:115]: I would suggest adding this:
```rust
        command(root.join("contrib/download_prerequisites")).current_dir(&root).run(builder);
```
before `configure`. It should make the build more self-contained.

File: src/bootstrap/src/core/build_steps/gcc.rs
Comment (GuillaumeGomez) [reply src/bootstrap/src/core/build_steps/gcc.rs:127]: It didn't in my case but better add a check for it.

**Comment by GuillaumeGomez** [comment]
Applied your suggestions @Kobzol. :)

**Comment by GuillaumeGomez** [comment]
Fixed merge conflicts.

@bors r=Kobzol

**Comment by Kobzol** [comment]
That being said, why is this step executed on CI..? :confused: That shouldn't happen.

File: src/bootstrap/src/core/build_steps/dist.rs
Comment (Kobzol) [inline src/bootstrap/src/core/build_steps/dist.rs:2276]: Oh this was kept in, we forgot to remove it. Please revert the changes to `dist` to make this PR only affect local builds.

**Comment by GuillaumeGomez** [comment]
Removed changes to `dist.rs`.

**Pull request merged**
